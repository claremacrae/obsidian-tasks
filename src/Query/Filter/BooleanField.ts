import { PostfixExpression, parse as boonParse } from 'boon-js';

import { parseFilter } from '../FilterParser';
import type { Task } from '../../Task';
import { Field } from './Field';
import { Filter, FilterOrErrorMessage } from './Filter';

export class BooleanField extends Field {
    private static readonly basicBooleanRegexp = /.*(AND|OR|XOR|NOT).*/g;
    private subFields: Record<string, Filter> = {};

    protected filterRegexp(): RegExp {
        return BooleanField.basicBooleanRegexp;
    }

    public createFilterOrErrorMessage(line: string): FilterOrErrorMessage {
        return this.parseLine(line);
    }

    protected fieldName(): string {
        return 'boolean query';
    }

    private parseLine(line: string): FilterOrErrorMessage {
        const result = new FilterOrErrorMessage();
        if (line.length === 0) {
            result.error = 'empty line';
            return result;
        }
        const preprocessed = this.preprocessExpression(line);
        try {
            const postfixExpression = boonParse(preprocessed);
            // Construct sub-field map, i.e. have subFields include a filter function for every
            // final token in the expression
            for (const token of postfixExpression) {
                if (token.name === 'IDENTIFIER' && token.value) {
                    if (!(token.value in this.subFields)) {
                        const parsedField = parseFilter(token.value);
                        if (parsedField === null) {
                            result.error = `couldn't parse sub-expression '${token.value}'`;
                            return result;
                        }
                        if (parsedField.error) {
                            result.error = `couldn't parse sub-expression '${token.value}': ${parsedField.error}`;
                            return result;
                        } else if (parsedField.filter) {
                            this.subFields[token.value] = parsedField.filter;
                        }
                    }
                }
            }
            result.filter = (task: Task) => {
                return this.filterTaskWithParsedQuery(task, postfixExpression);
            };
            return result;
        } catch (error) {
            const message =
                error instanceof Error ? error.message : 'unknown error type';
            result.error = `malformed boolean query -- ${message} (check the documentation for guidelines)`;
            return result;
        }
        return result;
    }

    private preprocessExpression(line: string): string {
        // Prepare the query to be processed by boon-js.
        // Boon doesn't process expression with spaces unless they are surrounded by quotes, so replace
        // (due today) by ("due today").
        return line.replace(/\(([^()]+)\)/g, '("$1")');
    }

    private filterTaskWithParsedQuery(
        task: Task,
        postfixExpression: PostfixExpression,
    ): boolean {
        const toBool = (s: string | undefined) => {
            return s === 'true';
        };
        const toString = (b: boolean) => {
            return b ? 'true' : 'false';
        };
        const booleanStack: string[] = [];
        for (const token of postfixExpression) {
            if (token.name === 'IDENTIFIER') {
                if (token.value == null) throw Error('null token value');
                // TODO document
                const filter = this.subFields[token.value];
                const result = filter(task);
                booleanStack.push(toString(result));
            } else if (token.name === 'OPERATOR') {
                if (token.value === 'NOT') {
                    const arg1 = toBool(booleanStack.pop());
                    booleanStack.push(toString(!arg1));
                } else if (token.value === 'OR') {
                    const arg1 = toBool(booleanStack.pop());
                    const arg2 = toBool(booleanStack.pop());
                    booleanStack.push(toString(arg1 || arg2));
                } else if (token.value === 'AND') {
                    const arg1 = toBool(booleanStack.pop());
                    const arg2 = toBool(booleanStack.pop());
                    booleanStack.push(toString(arg1 && arg2));
                } else {
                    throw Error('Unsuppoted operator' + token.value);
                }
            } else {
                throw Error('Unsupported token type:' + token);
            }
        }
        return toBool(booleanStack[0]);
    }
}
