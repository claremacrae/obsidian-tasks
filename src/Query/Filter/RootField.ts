import type { Task } from '../../Task';
import type { GrouperFunction } from '../Grouper';
import { TextField } from './TextField';

export class RootField extends TextField {
    public fieldName(): string {
        return 'root';
    }

    public value(task: Task): string {
        return task.file.root;
    }

    // TODO Replace this with use of TasksFile instead, after merge
    public static root(originalPath: string) {
        let path = originalPath.replace(/\\/g, '/');

        if (path.charAt(0) === '/') {
            path = path.substring(1);
        }

        const separatorIndex = path.indexOf('/');
        if (separatorIndex == -1) {
            return '/';
        }
        return path.substring(0, separatorIndex + 1);
    }

    public supportsGrouping(): boolean {
        return true;
    }

    public grouper(): GrouperFunction {
        return (task: Task) => {
            return [TextField.escapeMarkdownCharacters(this.value(task))];
        };
    }
}
