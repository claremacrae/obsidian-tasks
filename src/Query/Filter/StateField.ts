import { TaskStates } from '../../Status';
import type { Task } from '../../Task';
import { FilterInstructionsBasedField } from './FilterInstructionsBasedField';

export class StateField extends FilterInstructionsBasedField {
    constructor() {
        super();

        for (const state of TaskStates.states) {
            this._filters.add(
                `state is ${state.commandName}`,
                (task: Task) => task.originalStatusCharacter === state.statusCharacter,
            );
        }
    }

    public fieldName(): string {
        return 'state';
    }
}
