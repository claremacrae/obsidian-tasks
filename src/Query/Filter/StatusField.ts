import { Status } from '../../Status';
import type { Task } from '../../Task';
import { FilterInstructionsBasedField } from './FilterInstructionsBasedField';

export class StatusField extends FilterInstructionsBasedField {
    constructor() {
        super();

        this._filters.add('done', (task: Task) => task.status === Status.DONE);
        this._filters.add('not done', (task: Task) => task.status !== Status.DONE);

        this._filters.add('status is unchecked', (task: Task) => task.originalStatusCharacter === ' ');
        this._filters.add('status is important', (task: Task) => task.originalStatusCharacter === '!');
        this._filters.add('status is half-done', (task: Task) => task.originalStatusCharacter === '/');
        this._filters.add('status is cancelled', (task: Task) => task.originalStatusCharacter === '-');
        this._filters.add('status is doing', (task: Task) => task.originalStatusCharacter === 'd');
    }

    public fieldName(): string {
        return 'status';
    }
}
