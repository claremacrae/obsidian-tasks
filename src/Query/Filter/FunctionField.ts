import type { Task } from '../../Task';
import type { GrouperFunction } from '../Grouper';
import { Grouper } from '../Grouper';
import { Field } from './Field';
import { FilterOrErrorMessage } from './Filter';
import { HappensDateField } from './HappensDateField';
import { RootField } from './RootField';

export class FunctionField extends Field {
    createFilterOrErrorMessage(line: string): FilterOrErrorMessage {
        return FilterOrErrorMessage.fromError(line, 'Searching by custom function not yet implemented');
    }

    fieldName(): string {
        return 'function';
    }

    protected filterRegExp(): RegExp | null {
        return null;
    }

    // -----------------------------------------------------------------------------------------------------------------
    // Grouping
    // -----------------------------------------------------------------------------------------------------------------

    public supportsGrouping(): boolean {
        return true;
    }

    public createGrouperFromLine(line: string): Grouper | null {
        if (!this.supportsGrouping()) {
            return null;
        }

        const match = Field.getMatch(this.grouperRegExp(), line);
        if (match === null) {
            return null;
        }
        const args = match[1];

        return new Grouper('function', createGrouperFunctionFromLine(args));
    }

    protected grouperRegExp(): RegExp {
        if (!this.supportsGrouping()) {
            throw Error(`grouperRegExp() unimplemented for ${this.fieldNameSingular()}`);
        }

        return new RegExp(`^group by ${this.fieldNameSingularEscaped()} (.*)`);
    }

    public grouper(): GrouperFunction {
        return (_task: Task) => {
            return ['hello world'];
        };
    }
}

function parameterArguments(task: Task) {
    const paramsArgs: [string, any][] = [
        ['created', task.createdDate],
        ['description', task.description],
        ['done', task.doneDate],
        ['due', task.dueDate],
        ['filename', task.filename],
        ['happens', new HappensDateField().earliestDate(task)],
        ['header', task.precedingHeader],
        ['indentation', task.indentation],
        ['markdown', task.originalMarkdown],
        ['path', task.path.replace('.md', '')],
        ['priority', task.priority],
        ['recurrence', task.recurrence],
        ['root', new RootField().value(task)],
        ['scheduled', task.scheduledDate],
        ['scheduledDateIsInferred', task.scheduledDateIsInferred],
        ['start', task.startDate],
        ['status', task.status],
        ['t', task],
        ['tags', task.tags],
        ['task', task],
        ['urgency', task.urgency],
    ];
    return paramsArgs;
}

export type GroupingArg = string | null;

export function createGrouperFunctionFromLine(line: string): GrouperFunction {
    return (task: Task) => {
        return groupByFn(task, line);
    };
}

export function createGrouperFromLine(line: string): Grouper | null {
    return new Grouper('function', createGrouperFunctionFromLine(line));
}

export function groupByFn(task: Task, arg?: GroupingArg): string[] {
    const paramsArgs = parameterArguments(task);

    const params = paramsArgs.map(([p]) => p);
    const groupBy = arg && new Function(...params, `return ${arg}`);

    if (groupBy instanceof Function) {
        const args = paramsArgs.map(([_, a]) => a);
        const result = groupBy(...args);
        const group = typeof result === 'string' ? result : 'Error with group result';

        return [group];
    } else {
        return ['Error parsing group function'];
    }
}
