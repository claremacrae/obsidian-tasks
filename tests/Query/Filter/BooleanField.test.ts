/**
 * @jest-environment jsdom
 */
import moment from 'moment';
import { BooleanField } from '../../../src/Query/Filter/BooleanField';
import type { FilterOrErrorMessage } from '../../../src/Query/Filter/FilterOrErrorMessage';
import { TaskBuilder } from '../../TestingTools/TaskBuilder';
import { testFilter } from '../../TestingTools/FilterTestHelpers';
import { verifyBooleanExpressionExplanation, verifyBooleanExpressionPreprocessing } from './BooleanFieldVerify';

window.moment = moment;

function testWithDescription(filter: FilterOrErrorMessage, description: string, expected: boolean) {
    const builder = new TaskBuilder();
    testFilter(filter, builder.description(description), expected);
}

function createValidFilter(instruction: string) {
    // First make sure that BooleanField recognises the instruction as valid
    expect(new BooleanField().canCreateFilterForLine(instruction)).toEqual(true);

    return new BooleanField().createFilterOrErrorMessage(instruction);
}

describe('boolean query', () => {
    // These tests are intended to be really simple, so it is easy to reason about the correct behaviour of the code.
    describe('basic operators', () => {
        it('AND', () => {
            // Arrange
            const filter = createValidFilter(
                '"description includes d1" AND "description includes d2"', // Use "..." instead of (), for completeness
            );

            // Act, Assert
            testWithDescription(filter, 'neither', false);
            testWithDescription(filter, 'd1', false);
            testWithDescription(filter, 'd2', false);
            testWithDescription(filter, 'd1 d2', true);
        });

        it('OR', () => {
            // Arrange
            const filter = createValidFilter('(description includes d1) OR (description includes d2)');

            // Act, Assert
            testWithDescription(filter, 'neither', false);
            testWithDescription(filter, 'd1', true);
            testWithDescription(filter, 'd2', true);
            testWithDescription(filter, 'd1 d2', true);
        });

        it('XOR', () => {
            // Arrange
            const filter = createValidFilter('(description includes d1) XOR (description includes d2)');

            // Act, Assert
            testWithDescription(filter, 'neither', false);
            testWithDescription(filter, 'd1', true);
            testWithDescription(filter, 'd2', true);
            testWithDescription(filter, 'd1 d2', false);
        });

        it('NOT', () => {
            // Arrange
            const filter = createValidFilter('NOT (description includes d1)');

            // Act, Assert
            testWithDescription(filter, 'nothing', true);
            testWithDescription(filter, 'd1', false);
        });

        it('AND NOT', () => {
            // Arrange
            const filter = createValidFilter('(description includes d1) AND NOT (description includes d2)');

            // Act, Assert
            testWithDescription(filter, 'neither', false);
            testWithDescription(filter, 'd1', true);
            testWithDescription(filter, 'd2', false);
            testWithDescription(filter, 'd1 d2', false);
        });

        it('OR NOT', () => {
            // Arrange
            const filter = createValidFilter('(description includes d1) OR NOT (description includes d2)');

            // Act, Assert
            testWithDescription(filter, 'neither', true);
            testWithDescription(filter, 'd1', true);
            testWithDescription(filter, 'd2', false);
            testWithDescription(filter, 'd1 d2', true);
        });

        it('input components dirty with whitespaces', () => {
            const filter = createValidFilter(
                ' (description includes #context/location1) OR (description includes #context/location2 ) OR (  description includes #context/location3 ) OR   (  description includes #context/location4 )',
            );

            testWithDescription(filter, 'none', false);
            testWithDescription(filter, 'xxx #context/location1', true);
            testWithDescription(filter, 'xxx #context/location2', true);
            testWithDescription(filter, 'xxx #context/location3', true);
            testWithDescription(filter, 'xxx #context/location4', true);
        });

        it('should work with single filter in parentheses - via BooleanField', () => {
            // This tests the fix for
            //  https://github.com/obsidian-tasks-group/obsidian-tasks/discussions/975
            //  Support single filters surround by parentheses
            const filter = createValidFilter('(description includes #context/location1)');

            testWithDescription(filter, 'xxx #context/location1', true);
            testWithDescription(filter, 'xxx #context/location2', false);
        });

        it('should work with single filter in heavily nested parentheses - via BooleanField', () => {
            // Confirm that redundant () are ignored.
            const filter = createValidFilter('(((((description includes #context/location1)))))');

            testWithDescription(filter, 'xxx #context/location1', true);
            testWithDescription(filter, 'xxx #context/location2', false);
        });
    });

    describe('error cases - to show error messages', () => {
        it.each([
            [
                // force line break
                '',
                'empty line',
            ],

            // Incorrect numbers of filters/sub-expressions
            [
                'AND (description includes d1)',
                'malformed boolean query -- Invalid token (check the documentation for guidelines)',
            ],
            [
                'OR (description includes d1)',
                'malformed boolean query -- Invalid token (check the documentation for guidelines)',
            ],
            [
                // force line break
                'NOT (description blahblah d1)',
                "couldn't parse sub-expression 'description blahblah d1'",
            ],
            [
                'NOT (happens before blahblahblah)',
                "couldn't parse sub-expression 'happens before blahblahblah': do not understand happens date",
            ],

            // Missing spaces before operator
            [
                '(path includes a)AND (path includes b)',
                'malformed boolean query -- Unexpected character: A. A closing parenthesis should be followed by another closing parenthesis or whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a)AND NOT(path includes b)',
                'malformed boolean query -- Unexpected character: A. A closing parenthesis should be followed by another closing parenthesis or whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a)OR (path includes b)',
                'malformed boolean query -- Unexpected character: O. A closing parenthesis should be followed by another closing parenthesis or whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a)OR NOT (path includes b)',
                'malformed boolean query -- Unexpected character: O. A closing parenthesis should be followed by another closing parenthesis or whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a)XOR (path includes b)',
                'malformed boolean query -- Unexpected character: X. A closing parenthesis should be followed by another closing parenthesis or whitespace (check the documentation for guidelines)',
            ],

            // Missing spaces after operator
            [
                '(path includes a) AND(path includes b)',
                'malformed boolean query -- Unexpected character: (. Operators should be separated using whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a) AND NOT(path includes b)',
                'malformed boolean query -- Unexpected character: (. Operators should be separated using whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a) OR(path includes b)',
                'malformed boolean query -- Unexpected character: (. Operators should be separated using whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a) OR NOT(path includes b)',
                'malformed boolean query -- Unexpected character: (. Operators should be separated using whitespace (check the documentation for guidelines)',
            ],
            [
                '(path includes a) XOR(path includes b)',
                'malformed boolean query -- Unexpected character: (. Operators should be separated using whitespace (check the documentation for guidelines)',
            ],
            [
                'NOT(path includes b)',
                'malformed boolean query -- Unexpected character: (. Operators should be separated using whitespace (check the documentation for guidelines)',
            ],
        ])(
            'should report expected error message: on "%s" - expected "%s"',
            (instruction: string, expectedError: string) => {
                const filter = new BooleanField().createFilterOrErrorMessage(instruction);
                expect(filter.error).toStrictEqual(expectedError);
            },
        );

        // Have not managed to create instructions that trigger these errors:
        //      result.error = 'empty operator in boolean query';
        //      result.error = `unknown boolean operator '${token.value}'`;
    });
});

describe('boolean query - exhaustive tests', () => {
    it('preprocess', () => {
        verifyBooleanExpressionPreprocessing(BooleanField.preprocessExpression);
    });

    it('explain', () => {
        verifyBooleanExpressionExplanation();
    });
});
