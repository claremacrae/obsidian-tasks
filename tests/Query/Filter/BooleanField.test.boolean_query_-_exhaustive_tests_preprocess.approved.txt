Results of preprocessing boolean expressions




Input:
'"description includes d1" AND "description includes d2"'
=>
Result:
"\"description includes d1\" AND \"description includes d2\""

--------------------------------------------------------


Input:
'( (description includes a) AND (description includes b) ) AND (description includes c)'
=>
Result:
"( (\"description includes a\") AND (\"description includes b\") ) AND (\"description includes c\")"

--------------------------------------------------------


Input:
'( (description includes a) AND (description includes b) AND (description includes c) ) OR ( (description includes d) AND (description includes e) AND (description includes f) )'
=>
Result:
"( (\"description includes a\") AND (\"description includes b\") AND (\"description includes c\") ) OR ( (\"description includes d\") AND (\"description includes e\") AND (\"description includes f\") )"

--------------------------------------------------------


Input:
'( (description includes a) OR (description includes b) ) OR (description includes c)'
=>
Result:
"( (\"description includes a\") OR (\"description includes b\") ) OR (\"description includes c\")"

--------------------------------------------------------


Input:
'( (description includes a) OR (description includes b) OR (description includes c) ) AND ( (description includes d) OR (description includes e) OR (description includes f) )'
=>
Result:
"( (\"description includes a\") OR (\"description includes b\") OR (\"description includes c\") ) AND ( (\"description includes d\") OR (\"description includes e\") OR (\"description includes f\") )"

--------------------------------------------------------


Input:
'( (filter by function task.description.includes('a')) OR (filter by function task.description.includes('b')) OR (filter by function task.description.includes('c')) ) AND ( (filter by function task.description.includes('d')) OR (filter by function task.description.includes('e')) OR (filter by function task.description.includes('f')) )'
=>
Result:
"( (filter by function task.description.includes(\"'a'\")) OR (filter by function task.description.includes(\"'b'\")) OR (filter by function task.description.includes(\"'c'\")) ) AND ( (filter by function task.description.includes(\"'d'\")) OR (filter by function task.description.includes(\"'e'\")) OR (filter by function task.description.includes(\"'f'\")) )"

--------------------------------------------------------


Input:
'( description includes a ) AND ( (description includes b) AND (description includes c) )'
=>
Result:
"(\" description includes a \") AND ( (\"description includes b\") AND (\"description includes c\") )"

--------------------------------------------------------


Input:
'( description includes a ) OR ( (description includes b) OR (description includes c) )'
=>
Result:
"(\" description includes a \") OR ( (\"description includes b\") OR (\"description includes c\") )"

--------------------------------------------------------


Input:
'(((((description includes #context/location1)))))'
=>
Result:
"(((((\"description includes #context/location1\")))))"

--------------------------------------------------------


Input:
'(description includes #context/location1)'
=>
Result:
"(\"description includes #context/location1\")"

--------------------------------------------------------


Input:
'(description includes #context/location1) OR (description includes #context/location2 ) OR (  description includes #context/location3 ) OR   (  description includes #context/location4 )'
=>
Result:
"(\"description includes #context/location1\") OR (\"description includes #context/location2 \") OR (\"  description includes #context/location3 \") OR   (\"  description includes #context/location4 \")"

--------------------------------------------------------


Input:
'(description includes 1) AND (description includes 2) AND (description includes 3)'
=>
Result:
"(\"description includes 1\") AND (\"description includes 2\") AND (\"description includes 3\")"

--------------------------------------------------------


Input:
'(description includes 1) AND (description includes 2) AND (description includes 3) AND (description includes 4) AND (description includes 5) AND (description includes 6) AND (description includes 7) AND (description includes 8) AND (description includes 9)'
=>
Result:
"(\"description includes 1\") AND (\"description includes 2\") AND (\"description includes 3\") AND (\"description includes 4\") AND (\"description includes 5\") AND (\"description includes 6\") AND (\"description includes 7\") AND (\"description includes 8\") AND (\"description includes 9\")"

--------------------------------------------------------


Input:
'(description includes 1)   AND   (description includes 2)   AND   (description includes 3)   AND   (description includes 4)   AND   (description includes 5)   AND   (description includes 6)   AND   (description includes 7)   AND   (description includes 8)   AND   (description includes 9)'
=>
Result:
"(\"description includes 1\")   AND   (\"description includes 2\")   AND   (\"description includes 3\")   AND   (\"description includes 4\")   AND   (\"description includes 5\")   AND   (\"description includes 6\")   AND   (\"description includes 7\")   AND   (\"description includes 8\")   AND   (\"description includes 9\")"

--------------------------------------------------------


Input:
'(description includes d1) AND (priority medium)'
=>
Result:
"(\"description includes d1\") AND (\"priority medium\")"

--------------------------------------------------------


Input:
'(description includes d1) AND NOT (description includes d2)'
=>
Result:
"(\"description includes d1\") AND NOT (\"description includes d2\")"

--------------------------------------------------------


Input:
'(description includes d1) AND NOT (priority medium)'
=>
Result:
"(\"description includes d1\") AND NOT (\"priority medium\")"

--------------------------------------------------------


Input:
'(description includes d1) OR (description includes d2)'
=>
Result:
"(\"description includes d1\") OR (\"description includes d2\")"

--------------------------------------------------------


Input:
'(description includes d1) OR (description includes d2) OR (priority medium)'
=>
Result:
"(\"description includes d1\") OR (\"description includes d2\") OR (\"priority medium\")"

--------------------------------------------------------


Input:
'(description includes d1) OR (priority medium)'
=>
Result:
"(\"description includes d1\") OR (\"priority medium\")"

--------------------------------------------------------


Input:
'(description includes d1) OR NOT (description includes d2)'
=>
Result:
"(\"description includes d1\") OR NOT (\"description includes d2\")"

--------------------------------------------------------


Input:
'(description includes d1) OR NOT (priority medium)'
=>
Result:
"(\"description includes d1\") OR NOT (\"priority medium\")"

--------------------------------------------------------


Input:
'(description includes d1) XOR (description includes d2)'
=>
Result:
"(\"description includes d1\") XOR (\"description includes d2\")"

--------------------------------------------------------


Input:
'(description includes d1) XOR (priority medium)'
=>
Result:
"(\"description includes d1\") XOR (\"priority medium\")"

--------------------------------------------------------



Input:
'AND (description includes d1)'
=>
Result:
"AND (\"description includes d1\")"

--------------------------------------------------------


Input:
'NOT (description blahblah d1)'
=>
Result:
"NOT (\"description blahblah d1\")"

--------------------------------------------------------


Input:
'NOT (description includes d1)'
=>
Result:
"NOT (\"description includes d1\")"

--------------------------------------------------------


Input:
'NOT (happens before blahblahblah)'
=>
Result:
"NOT (\"happens before blahblahblah\")"

--------------------------------------------------------


Input:
'OR (description includes d1)'
=>
Result:
"OR (\"description includes d1\")"

--------------------------------------------------------


Input:
'(description includes SHOULD NOT BE RECOGNISED AS A BOOLEAN)AND(description includes BECAUSE THERE ARE NO SPACES AROUND THE 'AND' OPERATOR)'
=>
Result:
"(\"description includes SHOULD NOT BE RECOGNISED AS A BOOLEAN\")AND(\"description includes BECAUSE THERE ARE NO SPACES AROUND THE 'AND' OPERATOR\")"

--------------------------------------------------------



Input:
'(path includes (some example) OR (path includes )some example()'
=>
Result:
"(path includes (\"some example\") OR (\"path includes \")some example()"

--------------------------------------------------------


Input:
'(path includes (some example)) OR (path includes )some example()'
=>
Result:
"(path includes (\"some example\")) OR (\"path includes \")some example()"

--------------------------------------------------------


Input:
'(path includes )some example() OR (path includes (some example))'
=>
Result:
"(\"path includes \")some example() OR (path includes (\"some example\"))"

--------------------------------------------------------


Input:
'(path includes ()some example()) OR (path includes ((some example)))'
=>
Result:
"(path includes ()some example()) OR (path includes ((\"some example\")))"

--------------------------------------------------------



Input:
'( description regex matches /(buy|order|voucher|lakeland|purchase|\spresent)/i ) OR ( path includes Home/Shopping )'
=>
Result:
"( description regex matches /(\"buy|order|voucher|lakeland|purchase|\\spresent\")/i ) OR (\" path includes Home/Shopping \")"

--------------------------------------------------------


Input:
'( filter by function ! 'NON_TASK,CANCELLED'.includes(task.status.type) ) OR ( filter by function const date = task.due.moment; return date ? !date.isValid() : false; ) OR ( filter by function task.due.moment?.isSameOrBefore(moment(), 'day') || false ) OR ( filter by function task.urgency.toFixed(2) === 1.95.toFixed(2) ) OR ( filter by function (!task.isRecurring) && task.originalMarkdown.includes('🔁') ) OR ( filter by function task.file.path.toLocaleLowerCase() === 'TASKS RELEASES/4.1.0 RELEASE.MD'.toLocaleLowerCase() ) OR ( filter by function const taskDate = task.due.moment; const now = moment(); return taskDate?.isSame(now, 'day') || ( !taskDate && task.heading?.includes(now.format('YYYY-MM-DD')) ) || false ) OR ( filter by function const wanted = '#context/home'; return task.heading?.includes(wanted) || task.tags.find( (tag) => tag === wanted ) && true || false; )'
=>
Result:
"( filter by function ! 'NON_TASK,CANCELLED'.includes(\"task.status.type\") ) OR ( filter by function const date = task.due.moment; return date ? !date.isValid() : false; ) OR ( filter by function task.due.moment?.isSameOrBefore(moment(), 'day') || false ) OR ( filter by function task.urgency.toFixed(\"2\") === 1.95.toFixed(\"2\") ) OR ( filter by function (\"!task.isRecurring\") && task.originalMarkdown.includes(\"'🔁'\") ) OR ( filter by function task.file.path.toLocaleLowerCase() === 'TASKS RELEASES/4.1.0 RELEASE.MD'.toLocaleLowerCase() ) OR ( filter by function const taskDate = task.due.moment; const now = moment(); return taskDate?.isSame(\"now, 'day'\") || ( !taskDate && task.heading?.includes(now.format(\"'YYYY-MM-DD'\")) ) || false ) OR ( filter by function const wanted = '#context/home'; return task.heading?.includes(\"wanted\") || task.tags.find( (\"tag\") => tag === wanted ) && true || false; )"

--------------------------------------------------------


